### PR TITLE
Allow `subscribe` to NOT ack message according to handler response.

### PR DESCRIPTION
**Pull request by bertuss**

### CODE REVIEW

**Review by nicja — APPROVED** [review]
Nice. Is there no explicit way to nack a message? Also, please update the changelog and prepare a release.

**Comment by matthewgoslett** [comment]
@nicja @bertuss 

This breaks contract with the interface.

I'd recommend an accessor / config for this so that the `subscribe` signature stays as is.

File: src/GoogleCloudPubSubAdapter.php [hunk 1]
```diff
@@ -207,10 +208,12 @@ public function subscribe($channel, callable $handler)
                 if ($payload === 'unsubscribe') {
                     $isSubscriptionLoopActive = false;
                 } else {
-                    call_user_func($handler, $payload);
+                    $response = call_user_func($handler, $payload);
+                }
+                
+                if (!$checkResponse || isset($response) && $response) {
```
Comment (geyer-za) [inline src/GoogleCloudPubSubAdapter.php]: just for some compatibility.. what if we said `if($reponse !== false) {`?

Comment (bertuss) [reply src/GoogleCloudPubSubAdapter.php]: Compatibility? This is backwards compatible by default.
It retains the default expected behaviour by setting `$checkResponse` to false.
Only if it is set to true do we care about the response from the callback.
And then we also check that a response was returned and that that value is true or truthy.
```php
// default behaviour, retains backwards compatibility
>>> $checkResponse = false; 
=> false
>>> (!$checkResponse || isset($response) && $response)
=> true // always ack message

// we care about the response from the handler, but we don't pass a response
>>> $checkResponse = true; 
=> true 
>>> (!$checkResponse || isset($response) && $response)
=> false // no response, don't ack

// we care about the response from the handler, we return a response, true in this case
>>> $response = true;
=> true
>>> (!$checkResponse || isset($response) && $response)
=> true // we do ack

// we care about the response from the handler, we return a response, false
>>> $response = false;
=> false
>>> (!$checkResponse || isset($response) && $response)
=> false // handler failed or borked, don't ack, bail, bail, bail
```

File: src/GoogleCloudPubSubAdapter.php [hunk 2]
```diff
@@ -185,8 +185,9 @@ public function setMaxMessages($maxMessages)
      *
      * @param string $channel
      * @param callable $handler
+     * @param bool $checkResponse 
      */
-    public function subscribe($channel, callable $handler)
+    public function subscribe($channel, callable $handler, $checkResponse = false)
```
Comment (bertuss) [inline src/GoogleCloudPubSubAdapter.php]: `$checkResponse` is a bad name, suggestions?

Comment (nicja) [reply src/GoogleCloudPubSubAdapter.php]: As @matthewgoslett said we should rather try not to break the interface here..

**Review by nicja — CHANGES_REQUESTED** [review]
Needs to conform to `Superbalist\PubSub\PubSubAdapterInterface`

File: src/GoogleCloudPubSubAdapter.php (hunk 2 above)
Comment (bertuss) [reply src/GoogleCloudPubSubAdapter.php]: Done, however, `checkResponse` is still a bad name, any ideas?

Comment (bertuss) [reply src/GoogleCloudPubSubAdapter.php]: Thanks for keeping me in check @matthewgoslett ❤️

**Review by nicja — APPROVED** [review]
`alwaysAck`? Naming things is hard

**Comment by bertuss** [comment]
`alwaysAck` 👍 at least it does what it says, will amend. Dankie

**Comment by nicja** [comment]
nice! :+1:

**Comment by matthewgoslett** [comment]
Just a suggestion, but why not wrap calling the callback/handler in a
try/catch.  If any exception occurs from the handler, message is nacked,
otherwise message is acked.
This would also stop the subscriber dying, and restarting, every time an
app exception occurs.

**Comment by bertuss** [comment]
@matthewgoslett Good idea. How do we feel about libraries logging on those caught exceptions?
Inject a `Psr\Log\LoggerInterface` logger? Defaulting to stdout if no logger provided?

**Pull request closed without merging**